python/tvm/relay/quantize/quantize.py

```diff
@@ -373,3 +738,135 @@ def quantize(graph, params=None, dataset=None):
+    batch_size = 1
+    oshape = (batch_size, num_classes)
+    out_arr = tvm.nd.empty(oshape, "float32")
+    # setup evaluaiton metric
+    acc_top1 = mx.metric.Accuracy()
+    acc_top5 = mx.metric.TopKAccuracy(5)
+    val_data.reset()
+    acc_top1.reset()
+    acc_top5.reset()
+    # execute
+
+    output_collection = [None]*(m.get_num_outputs() - 1)
+    for i, batch in enumerate(val_data):
+        data, label = batch_fn(batch, [mx.cpu(0)])
+        m.run(data=data[0].asnumpy())
+        m.run(data=data[0].asnumpy(), **scales)
+        m.get_output(0, out_arr)
+        for o in range(0, len(output_collection)):
+            if output_collection[o] is None:
+                output_collection[o] = m.get_output(o+1).asnumpy()
+            else:
+                output_collection[o] = np.concatenate((output_collection[o], m.get_output(o+1).asnumpy()))
+        acc_top1.update(label, [mx.nd.array(out_arr.asnumpy())])
+        acc_top5.update(label, [mx.nd.array(out_arr.asnumpy())])
+        _, top1 = acc_top1.get()
+        _, top5 = acc_top5.get()
+
+        if not (i + 1) % log_iter:
+            nsamples = (i + 1) * batch_size
+            print('[{0:d} samples] evaluation: acc-top1={1:f} acc-top5={2:f}'.format(nsamples, top1, top5))
+
+        if (i+1)*batch_size >= early_stopping:
+            return top1, output_collection
+
+#def autoquantize(graph_callback, tr_data, tr_batch_fn, granularity='layer'):
+def autoquantize(graph, params, tr_data, tr_batch_fn, granularity='layer'):
+
+    import tvm
+    import copy
+    from tvm import relay
+    from tvm.relay import ir_pass
+
+    #graph, params = graph_callback()
+
+    graph = optimize(graph, params)
+    with qconfig(skip_k_conv=0,
+                 passthrough_bound=int(-1),
+                 nbit_input=8,
+                 nbit_weight=8,
+                 global_scale=8.0,
+                 dtype_input='int8',
+                 dtype_weight='int8',
+                 dtype_activation='int32',
+                 store_lowbit_output=True,
+                 debug_enabled_ops=None,
+                 granularity=granularity):
+        layout_map = tag_layout(graph)
+        graph = annotate(graph, layout_map)
+        annotated = copy.deepcopy(graph)
+        graph = ir_pass.infer_type(graph)
+        graph, profile_data = calibrate(graph, profile_mode=True, scales=None)
+        
+        free_vars = list(ir_pass.free_vars(graph))
+        graph = relay.Function(list(graph.params) + free_vars,
+                                graph.body, graph.ret_type,
+                                graph.type_params, graph.attrs)
+        additional_outputs = list()
+        metadata = list()
+        for hint, data, granularity, layout in profile_data:
+            additional_outputs.append(data)
+            metadata.append((hint, granularity, layout))
+        graph = relay.Function(graph.params,
+                                relay.expr.Tuple([graph.body]+additional_outputs))
```

Comment:
Why should we add the final ouptut(`graph.body`) to graph extra ? I think `graph.body` is included in `additional_outputs`, which contains all quantization op and quantize_kind != WEIGHT. Am i misunderstand ?

Reply:
This is not used for the final result, just to profile intermediate activations during calibration.

Reply:
I mean `additional_outputs` includes all internal outputs. because the `profile_data` collects all internal act outputs by  `visit_func`. `graph.body` only contains the final output. And `additional_outputs` also contains the final output. Is there  a redundancy about `the final output` ?